File: AIReaderKindle/src/Domain/AI/SearchTool.cpp
#include "SearchTool.hpp"

#include "ToolSchema.hpp"

namespace SearchTool {

const char* const toolName = "search_book";

Json tool() {
    return ToolSchema::function(
        toolName,
        "Ищет слово или фразу в тексте книги, которую читает пользователь (и в других книгах той же серии), "
        "и возвращает отрывки, где она встречается, — только до места, до которого читатель дочитал.",
        "query",
        "Слово или короткая фраза, которую нужно найти в книге.");
}

std::string query(const std::string& arguments) {
    return ToolSchema::argument(arguments, "query", "");
}

std::string summary(const std::string& query, const std::vector<SearchHit>& hits, bool severalBooks) {
    if (hits.empty()) return "No passage read so far contains “" + query + "”.";
    std::string text = "Passages with “" + query + "”:\n";
    for (size_t i = 0; i < hits.size(); ++i) {
        const SearchHit& hit = hits[i];
        text += std::to_string(i + 1) + ". [";
        if (severalBooks) text += hit.bookTitle + ", ";
        text += "chapter " + std::to_string(hit.chapter + 1) + "] " + hit.excerpt + "\n";
    }
    return text;
}

}  // namespace SearchTool

File: AIReader/AIReader/Domain/AI/SearchTool.swift
import Foundation

/// The tool that lets the model read the book: `search_book` finds a phrase
/// in the book being read, and in the other books of its group.
enum SearchTool {
    static let toolName = "search_book"

    /// How many passages a call returns to the model.
    static let passageLimit = 12

    static let tool: [String: Any] = ToolSchema.function(
        name: toolName,
        description: "Ищет слово или фразу в тексте книги, которую читает пользователь (и в других книгах той же серии), "
            + "и возвращает отрывки, где она встречается, — только до места, до которого читатель дочитал.",
        argument: "query",
        argumentDescription: "Слово или короткая фраза, которую нужно найти в книге."
    )

    static func query(in arguments: String) -> String {
        ToolSchema.argument("query", in: arguments) ?? ""
    }

    /// The passages as the model reads them: numbered, each with its book when
    /// several are searched, and its chapter.
    static func summary(query: String, hits: [SearchHit], severalBooks: Bool) -> String {
        guard !hits.isEmpty else { return "No passage read so far contains “\(query)”." }
        var text = "Passages with “\(query)”:\n"
        for (index, hit) in hits.enumerated() {
            text += "\(index + 1). ["
            if severalBooks { text += "\(hit.bookTitle), " }
            text += "chapter \(hit.chapter + 1)] \(hit.excerpt)\n"
        }
        return text
    }
}
